tests/downloadwithretry_test.sh

```diff
@@ -0,0 +1,283 @@
+  assertEquals "incorrect result code" "1" "$?" || return
+  echo "$DOWNLOAD_OUTPUT" | grep "FAILED: could not retrieve ${bigFileName}" | wc -l | grep "^\s*2\s*$" 2>${stderr} 1>${stdout}
+  assertEquals "there should be ${EXPECTED_RETRIES} failed statements" "0" "$?" || return
+  echo "$DOWNLOAD_OUTPUT" | grep "ERROR: could not retrieve ${TARGET_URL}" 2>${stderr} 1>${stdout}
+  assertEquals "there should be an ERROR statement" "0" "$?" || return
+}
+
+testIfFileExists() {
+  # fail "DISABLED" || return
+  local DOWNLOAD_OUTPUT
+
+  assertFalse "${smallFileName} should not exist" "[ -f ${smallFileName} ]" || return
+  touch ${smallFileName}
+
+  echo -ne "\tTrying to download ${smallFileName}...\n"
+  DOWNLOAD_OUTPUT=$(downloadWithRetry ${smallFileUrl})
+  assertEquals "incorrect result code" "2" "$?" || return
+  echo "$DOWNLOAD_OUTPUT" | grep "ERROR: ${smallFileName} already exists." 2>${stderr} 1>${stdout}
+  assertEquals "there should be an ERROR statement" "0" "$?" || return
+}
+
+testSuccessfulDownload() {
+  # fail "DISABLED" || return
+  local DOWNLOAD_OUTPUT
+
+  assertFalse "${smallFileName} should not exist" "[ -f ${smallFileName} ]" || return
+  assertFalse "${smallFileMD5Name} should not exist" "[ -f ${smallFileMD5Name} ]" || return
+
+  echo -ne "\tTrying to download ${smallFileName}...\n"
+  DOWNLOAD_OUTPUT=$(downloadWithRetry ${smallFileUrl})
+  assertEquals "incorrect result code" "0" "$?" || return
+  echo "$DOWNLOAD_OUTPUT" | sed '/^$/d' | wc -l | grep "^\s*0\s*$" 2>${stderr} 1>${stdout}
+  assertEquals "there should be no output generated" "0" "$?" || return
+
+  echo -ne "\tTrying to download ${smallFileMD5Name}...\n"
+  DOWNLOAD_OUTPUT=$(downloadWithRetry ${smallFileMD5Url})
+  assertEquals "incorrect result code" "0" "$?" || return
+  echo "$DOWNLOAD_OUTPUT" | sed '/^$/d' | wc -l | grep "^\s*0\s*$" 2>${stderr} 1>${stdout}
+  assertEquals "there should be no output generated" "0" "$?" || return
+
+  echo -ne "\tChecking download integrity...\n"
+  echo -ne "\t"
+  md5sum -c ${smallFileMD5Name}
+  assertEquals "incorrect result code" "0" "$?" || return
+}
+
+testSuccessfulRetriedDownload() {
+  # fail "DISABLED" || return
+  local dlPID
+  local DOWNLOAD_OUTPUT
+
+  assertFalse "${smallFileName} should not exist" "[ -f ${smallFileName} ]" || return
+  assertFalse "${smallFileMD5Name} should not exist" "[ -f ${smallFileMD5Name} ]" || return
+
+  echo -ne "\tTrying to download ${smallFileName}...\n"
+  DOWNLOAD_OUTPUT=$(downloadWithRetry ${smallFileUrl} 2>${stderr} 1>${stdout}&)
+  dlPID=$(pgrep curl)
+  assertEquals "curl should be running" "0" "$?" || return
+  echo -ne "\tCurl is runnning, killing it and waiting 2 seconds\n"
+  kill -9 ${dlPID}
+  sleep 2
+  pgrep curl
+  assertNotEquals "curl should not be running" "0" "$?" || return
+  echo -ne "\tCurl is not runnning, waiting 10 seconds for download to finish...\n"
+  sleep 10
+  pgrep curl
+  assertNotEquals "curl should have finished downloading" "0" "$?" || return
```

Comment:
This test keeps failing here:
```
testSuccessfulRetriedDownload
        Trying to download nuxeo-shell-5.3.1.zip...
        Curl is runnning, killing it and waiting 2 seconds
        Curl is not runnning, waiting 10 seconds for download to finish...
2475
ASSERT:curl should have finished downloading expected not same```
The `ps` command indeed show me a curl process running.